File: ifif/python/S3_1913.py
import sys


sys.stdin = open("input.txt", "r")

N = int(input())
target = int(input())

direction = [(1, 0), (0, 1), (-1, 0), (0, -1)]
arr = [[0 for _ in range(N)] for _ in range(N)]
arr[0][0] = N**2
x, y, idx = 0, 0, 0
n = N**2-1

def snail(x, y, n, N, idx):
    d = direction[idx]
    dx = x + d[0]
    dy = y + d[1]
    if 0<=dx<N and 0<=dy<N and arr[dx][dy]==0:
        x, y = dx, dy
        arr[x][y] = n
        n -= 1
    else:
        idx += 1
    return x, y, n, idx

while n>=1:
    idx = idx%4
    x, y, n, idx = snail(x, y, n, N, idx)
target_x = 0
target_y = 0
for row in range(N):
    if target in arr[row]:
        target_x = row
        target_y = arr[row].index(target)
    print(*arr[row])
print(target_x+1, target_y+1)

File: ifif/java/S3_1913.java
import java.io.*;
import java.util.Scanner;

public class S3_1913 {
    public static void main(String[] args) throws IOException {
        System.setIn(new FileInputStream("input.txt"));
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int N = Integer.parseInt(reader.readLine());
        int target = Integer.parseInt(reader.readLine());

        int num = N*N;
        int[][] arr = new int[N][N];
        int[][] direction = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
        int idx = 0;
        int x = -1;
        int y = 0;
        int[] target_idx = {0, 0};
        while (num >=1) {
            int dx = x + direction[idx][0];
            int dy = y + direction[idx][1];
            if ((0<=dx && dx<N) && (0<=dy && dy<N) && arr[dx][dy]==0) {
                x = dx;
                y = dy;
                arr[x][y] = num;
                if (num == target) {
                    target_idx[0] = x+1;
                    target_idx[1] = y+1;
                }
                num--;
            }
            else {
                idx++;
                idx %= 4;
            }
        }
        for (int[] ar : arr) {
            for (int a : ar) {
                sb.append(a).append(" ");
            }
            sb.append("\n");
        }

        sb.append(target_idx[0]).append(" ").append(target_idx[1]);
        System.out.println(sb);
    }
}
